From 1575633c6589de5aeab53e50531b5fdbab4f21ec Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@manjaro.org>
Date: Sun, 26 May 2024 16:26:53 +0200
Subject: [PATCH 4/5] arm: sunxi: add support for OtterCastAudioV2

The OtterCastAudioV2 is a single board computer based on the
Sochip S3 SoC optimized for audio streaming applications. It features
an overall small footprint and an on-board high quality ADC and DAC for
audio out- and input.

Specifications:
 - SoC: Sochip S3 SiP
 - RAM: 128MiB DDR3 (in-package)
 - Storage: microSD card
 - Power: 5V USB-C
 - Ethernet: 100Mbit/s Ethernet
 - WLAN & BT: AP6236 WLAN and Bluetooth module
 - Audio: SSM2603 I2S codec

Signed-off-by: Tobias Schramm <t.schramm@manjaro.org>
---
 arch/arm/dts/Makefile                        |   1 +
 arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts | 123 +++++++++++++++++++
 configs/ottercast_audio_v2_defconfig         |  14 +++
 3 files changed, 138 insertions(+)
 create mode 100644 arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
 create mode 100644 configs/ottercast_audio_v2_defconfig

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index b102ffb5f6..9eaf4b8b8e 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -810,6 +810,7 @@ dtb-$(CONFIG_MACH_SUN8I_R40) += \
 	sun8i-v40-bananapi-m2-berry.dtb
 dtb-$(CONFIG_MACH_SUN8I_V3S) += \
 	sun8i-s3-elimo-initium.dtb \
+	sun8i-s3-ottercast-audio-v2.dtb \
 	sun8i-s3-pinecube.dtb \
 	sun8i-v3-sl631-imx179.dtb \
 	sun8i-v3s-licheepi-zero.dtb
diff --git a/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts b/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
new file mode 100644
index 0000000000..63a13ee2b2
--- /dev/null
+++ b/arch/arm/dts/sun8i-s3-ottercast-audio-v2.dts
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR X11)
+/*
+ * Copyright 2021 Tobias Schramm <t.schramm@t-sys.eu>
+ */
+
+/dts-v1/;
+#include "sun8i-v3.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "OtterCast Audio V2";
+	compatible = "ottercast,ottercast-audio-v2", "allwinner,sun8i-s3";
+
+	aliases {
+		serial0 = &uart1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led1 {
+			label = "led:usr";
+			gpios = <&pio 1 4 GPIO_ACTIVE_HIGH>; /* PB4 */
+		};
+
+		led2 {
+			label = "led:pwr";
+			gpios = <&pio 1 5 GPIO_ACTIVE_HIGH>; /* PB5 */
+			default-state = "on";
+		};
+	};
+
+	dcdc_1v1: dcdc-1v1 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "dcdc-3v3";
+		regulator-min-microvolt = <1100000>;
+		regulator-max-microvolt = <1100000>;
+	};
+
+	dcdc_3v3: dcdc-3v3 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "dcdc-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+	};
+
+	ldo_rtc: ldo-rtc {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "ldo-rtc";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+	};
+
+	ldo_3v0: ldo-3v0 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "ldo-3v0";
+		regulator-min-microvolt = <3000000>;
+		regulator-max-microvolt = <3000000>;
+	};
+
+	wifi_pwrseq: wifi_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&pio 4 19 GPIO_ACTIVE_LOW>; /* PE19 WL-PMU-EN */
+		post-power-on-delay-ms = <200>;
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	status = "okay";
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&mmc0 {
+	vmmc-supply = <&dcdc_3v3>;
+	bus-width = <4>;
+	cd-gpios = <&pio 5 6 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&dcdc_3v3>;
+	vqmmc-supply = <&dcdc_3v3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	vcc-pd-supply = <&dcdc_3v3>;
+	vcc-pe-supply = <&dcdc_3v3>;
+};
+
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pg_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&pio {
+	uart1_pg_pins: uart1-pg-pins {
+		pins = "PG6", "PG7";
+		function = "uart1";
+	};
+};
diff --git a/configs/ottercast_audio_v2_defconfig b/configs/ottercast_audio_v2_defconfig
new file mode 100644
index 0000000000..3ba1ffdbc1
--- /dev/null
+++ b/configs/ottercast_audio_v2_defconfig
@@ -0,0 +1,14 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-s3-ottercast-audio-v2"
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_V3S=y
+CONFIG_SUNXI_DRAM_DDR3_1333=y
+CONFIG_DRAM_CLK=504
+CONFIG_DRAM_ODT_EN=y
+CONFIG_I2C1_ENABLE=y
+# CONFIG_HAS_ARMV7_SECURE_BASE is not set
+CONFIG_SYS_I2C_MVTWSI=y
+CONFIG_DM_ETH_PHY=y
+CONFIG_SUN8I_EMAC=y
+CONFIG_CONS_INDEX=2
-- 
2.30.2

